## Variáveis

Até o momento, temos verificado o comportamento de valores e aprendido um pouco sobre os tipo de dados em Python. No entanto, todos os valores que utilizamos nas operações foram "perdidos" após obtermos um resultado. Observe:

In [2]:
3.2 + 7

10.2

Na operação aritmética acima, utilizamos dois valores (3.2 e 7) e visualizamos um terceiro valor (10.2), porém nossa interação com esses valores acaba aí. Não podemos "reutilizá-los". Essa capacidade de reutilização de dados será muito útil no desenvolvimento de programas e, no caso da linguagem Python, podemos fazer isso por meio de variáveis.

---
### Modelo de Memória e Variáveis

Antes de explicarmos o que é uma variável, vamos entender um pouco como é que o computador lida com esses valores quando escrevemos expressões como 3.2 + 7.

Cada um dos valores apresentados precisam existir na memória do computador para que possam ser manipulados. Então, ao digitar 3.2 + 7, é necessário que o computador aloque (reserve) um espaço de memória para representar o número 3.2 e um outro espaço de memória para o número 7.

Ao interpretar a operação de soma, identificada pelo operador **+**, o computador aloca um terceiro espaço de memória onde ele escreve o resultado da operação (10.2). Cada espaço de memória possui uma numeração, um endereço que o identifica.

Uma variável é uma forma da gente guardar um endereço de memória para poder acessar novamente o valor que estiver escrito neste endereço. Observe:

In [7]:
num = 3.2

A instrução acima é uma **instrução de atribuição**. Dizemos que **o valor 3.2 foi atribuído à variável num**. O resultado dessa instrução é o seguinte: o valor 3.2 é escrito em algum pedaço da memória do computador, que tem um determinado endereço; a variável **num** passa, então, a ser uma referência para o endereço de memória onde 3.2 está escrito. Assim, se precisarmos do valor 3.2 novamente, podemos usar a variável **num**. Veja:

In [8]:
num

3.2

Quando utilizamos num em uma expressão ou como um valor, o Python verifica qual o endereço de memória que num contém e, então, resgata o valor que estiver escrito naquele endereço. Assim, podemos operar:

In [9]:
num + 7

10.2

E continuamos com o valor 3.2 no endereço referenciado por num.

In [10]:
num

3.2

In [ ]:
Podemos atribuir um novo valor a uma variável já existente:

In [14]:
num = 42

Nesse caso, o número 42 é escrito em um outro endereço de memória e a variável num passa a ser uma referência para esse novo endereço de memória. O endereço de memória que num referenciava, onde havia o valor 3.2, deixa de ser parte do seu programa e é liberado automaticamnte de volta para outros usos pelo seu computador (a menos que haja alguma outra variável no seu programa que seja referência para aquele endereço). Por exemplo:

In [15]:
copia = num
# copia e num são referencias para o mesmo endereço de memória, onde há o valor 42

In [19]:
num = "biscoito"
# num torna-se uma referência para o endereço onde a string "biscoito" foi escrita. O endereço onde há 42 
# ainda faz parte do nosso programa, pois a variável "copia" o contém.

In [20]:
copia

42

In [21]:
num

'biscoito'

O site [Python Tutor](http://pythontutor.com) é uma ótima referência para se observar um modelo visual que representa esse processo de alocação de memória e criação de variáveis. 

#### Nomenclatura de variáveis

Podemos utilizar praticamente qualquer nome para variáveis. Há porém algumas pequenas restrições:

1. Variáveis não podem ter nomes iguais às palavras reservadas da linguagem Python (veremos várias delas nas seções seguintes, como if, else, while, for etc)

2. Variáveis não podem começar por números, porém podem ter 